lib/rucio/core/replica.py

```diff
@@ -569,9 +569,21 @@ def _resolve_dids(dids, unavailable, ignore_availability, all_states, session):
                                    models.DataIdentifier.name == did['name']))
 
     if did_clause:
-        for scope, name, did_type in session.query(models.DataIdentifier.scope,
-                                                   models.DataIdentifier.name,
-                                                   models.DataIdentifier.did_type).with_hint(models.DataIdentifier, "INDEX(DIDS DIDS_PK)", 'oracle').filter(or_(*did_clause)):
+        for scope, name, did_type, constituent in session.query(models.DataIdentifier.scope,
+                                                                models.DataIdentifier.name,
+                                                                models.DataIdentifier.did_type,
+                                                                models.DataIdentifier.constituent)\
+                                                         .with_hint(models.DataIdentifier, "INDEX(DIDS DIDS_PK)", 'oracle')\
+                                                         .filter(or_(*did_clause)):
+            if constituent:
+                # file is a constituent, resolve to any parent archive
+                archive = session.query(models.ConstituentAssociation.scope,
+                                        models.ConstituentAssociation.name)\
+                                 .filter(models.ConstituentAssociation.child_scope == scope,
+                                         models.ConstituentAssociation.child_name == name).one()
+                constituents['%s:%s' % (scope, name)] = {'scope': archive[0],
+                                                         'name': archive[1]}
```

Comment:
I think this is not enough. You have to get the archive somehow into the `file_clause`.
Otherwise the replicas are resolved in line 732, but essentially for most constituent files there will be no replica resolved, as there is none (it is only part of the archive).
So the whole resolving has to be done with the archive (pointing to the constituent)